sourced/ml/algorithms/uast_struct_to_bag.py

```diff
@@ -87,103 +100,77 @@ def __init__(self, p_explore_neighborhood, q_leave_neighborhood, n_walks, n_step
         so there are only 2 options with unnormalized probabilities 1/p & 1/q
         :param walk: list of visited nodes
         :return: next node to visit
         """
-        # notation from article - t - node from previous step, v - last node
-        v = walk[-1]
+        last_node = walk[-1]
 
-        if len(walk) == 1 and len(v.children) > 0:
-            return random.choice(v.children)
-        elif len(v.children) == 0:
-            return v
+        if len(walk) == 1 and len(last_node.children) > 0:
+            return random.choice(last_node.children)
+        elif len(last_node.children) == 0:
+            return last_node
 
-        t = walk[-2]
         threshold = (1 / self.p_explore_neighborhood)
-        threshold /= ((1 / self.p_explore_neighborhood) +
-                      len(v.children) / self.q_leave_neighborhood)
+        threshold /= (threshold + len(last_node.children) / self.q_leave_neighborhood)
 
         if random.random() <= threshold:
-            return t
+            return walk[-2]
 
-        options = []
-        if v.parent is not None:
-            options.append(v.parent)
-        options.extend(v.children)
-        return random.choice(options)
+        return random.choice(last_node.neighbours)
 
 
-class UastRandomWalk2Bag(UastStructure2BagBase):
+class UastRandomWalk2Bag(Uast2StructBagBase):
     def __init__(self, p_explore_neighborhood=0.5, q_leave_neighborhood=0.5, n_walks=5, n_steps=19,
                  stride=1, seq_len=(5, 6), seed=42):
-        self.random_walker = Uast2RandomWalks(p_explore_neighborhood=p_explore_neighborhood,
-                                              q_leave_neighborhood=q_leave_neighborhood,
-                                              n_walks=n_walks, n_steps=n_steps)
-        self.stride = stride
-
-        if not isinstance(seq_len, (int, tuple, list)):
-            raise TypeError("Unexpected type of seq_len: %s" % type(seq_len))
-
-        self.seq_len = seq_len
-        self.seed = seed
-
-    def uast_to_bag(self, uast):
-        if self.seed is not None:
-            random.seed(self.seed)
+        super().__init__(stride, seq_len)
+        self.uast2walks = Uast2RandomWalks(p_explore_neighborhood=p_explore_neighborhood,
+                                           q_leave_neighborhood=q_leave_neighborhood,
+                                           n_walks=n_walks, n_steps=n_steps, seed=seed)
 
+    def __call__(self, uast):
         bag = defaultdict(int)
-        walks = self.random_walker.uast2walks(uast)
-        if isinstance(self.seq_len, int):
-            seq_lens = [self.seq_len]
-        else:
-            seq_lens = self.seq_len
-        for walk in walks:
-            for seq_len in seq_lens:
-                for i in range(0, len(walk) - seq_len, self.stride):
+        for walk in self.uast2walks(uast):
+            for seq_len in self._seq_lens:
+                for i in range(0, len(walk) - seq_len, self._stride):
```

Comment:
`len(walk) - seq_len` -> `len(walk) - seq_len + 1` not to lose last element - bug in initial implementation